### ISSUE TITLE
difference between "Keep Glyphs" vs "Export Glyphs" custom params?

### ISSUE THREAD

**Issue by anthrotype** [googlefonts/glyphsLib#295]
fontmake currently uses a combination of "com.schriftgestaltung.Glyphs.Export" (bool, inside the GLIF lib) and "com.schriftgestaltung.Keep Glyphs" (list of strings, in the global UFO lib) in order to determine the list of final glyphs to keep in the generated font (currently using the fonttools subsetter after the fact, but I'd like to change this, anyway that's another issue..).

https://github.com/googlei18n/fontmake/blob/6cdf44ea38d17f9bb14fa65c1c9bc6ea82e3b89b/Lib/fontmake/font_project.py#L300-L304

In the Glyphs handbook I see two very similar custom parameters and I can't understand what is the difference between the two.

Can anyone please enlighten me? Thanks

/cc @schriftgestalt @moyogo @belluzj 

```
Keep Glyphs (list) List of glyphs that will be kept in the exported
  font. All other glyphs will be discarded, and kerning and
  automatic feature code will be updated accordingly. Works as
  the opposite of the Remove Glyphs parameter, useful for
  webfont subsetting in order to achieve smaller file sizes.
  Remove Glyphs and Keep Glyphs are mutually exclusive.

Export Glyphs (list) Exports all glyphs listed, regardless of whether
  the glyph was set to export or not.
```

**Comment by anthrotype** [comment]
is the key difference that "regardless of whether the glyph was set to export or not"? because by just reading the description of "Keep Glyphs" it's definitely not clear the connection with the glyph "export" flag...

**Comment by anthrotype** [comment]
and do we need to support that "Export Glyphs" somehow?

(I wonder why one needs _three_ ways to specify a font's subset)

**Comment by schriftgestalt** [comment]
The .exports key is ignored when the glyph is mentioned in a ‘Keep Glyphs’ parameter.

**Comment by anthrotype** [comment]
wait.. you actually mean the .exports key is ignored when the glyph is mentioned in _Export Glyphs_?
that's what I understand from "regardless of whether the glyph was set to export or not"... :confused:

**Comment by schriftgestalt** [comment]
I just had another look. There are two different parameters. 

- Keep Glyphs: only glyphs mentioned are exported. All other are not.
- Export Glyphs: the glyphs in this list will be set to export. No influence on the other glyphs

**Comment by schriftgestalt** [comment]
I think you are doing it right. The `Keep Glyphs` parameter is optional, and if not there, the export setting per glyph is respected. 

There is one more parameter in that group: `Remove Glyphs`. That is doing the opposite of `Export Glyphs` by disabling the .export setting of the mentioned glyphs.

**Comment by anthrotype** [comment]
> I think you are doing it right.

No, this is what fontmake is doing: it first checks if the keep_glyphs list is not empty; then it checks if the glyph name is present in there; if not, it skips the glyph and continues the loop; (so far so good). However, if the keep_glyphs list is either missing or is empty, or it is non-empty and the glyph name is contained in there, fontmake goes on checking the individual glyph exports flag: if the latter is present and set to False, fontmake skips the glyph, despite the global keep_glyphs list said the contrary.

this is the code:

https://github.com/googlei18n/fontmake/blob/6cdf44ea38d17f9bb14fa65c1c9bc6ea82e3b89b/Lib/fontmake/font_project.py#L326-L329

**Comment by schriftgestalt** [comment]
I do is

**Comment by madig** [comment]
So, "Keep Glyphs" overrides everything, then the export flag on a glyph is honored except if the glyph name is in "Export Glyphs" or not honored if it is in "Remove Glyphs"?

**Comment by anthrotype** [comment]
I don't think we need to support all _four_ ways to subset glyphs. It's already complicated like this, with global "Keep Glyphs" and per-glyph "Export" flags. Let's keep it simple
